qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -0,0 +1,363 @@
+
+            try:
+                _, successors = next(node_successor_generator)
+                successors = list(filter(lambda x: x.type == 'op' and len(x.qargs) == 2,
+                                         successors))
+            except StopIteration:
+                node_lookahead_exhausted[i] = True
+                continue
+
+            successors = iter(successors)
+            while len(extended_set) < EXTENDED_SET_SIZE:
+                try:
+                    extended_set.add(next(successors))
+                except StopIteration:
+                    break
+
+        return extended_set
+
+    def _obtain_swaps(self, front_layer, current_layout):
+        """Return a set of candidate swaps that affect qubits in front_layer.
+
+        For each virtual qubit in front_layer, find its current location
+        on hardware and the physical qubits in that neighborhood. Every SWAP
+        on virtual qubits that corresponds to one of those physical couplings
+        is a candidate SWAP.
+
+        Candidate swaps are sorted so SWAP(i,j) and SWAP(j,i) are not duplicated.
+        """
+        candidate_swaps = set()
+        for node in front_layer:
+            for virtual in node.qargs:
+                physical = current_layout[virtual]
+                for neighbor in self.coupling_map.neighbors(physical):
+                    virtual_neighbor = current_layout[neighbor]
+                    swap = sorted([virtual, virtual_neighbor],
+                                  key=lambda q: (q.register.name, q.index))
+                    candidate_swaps.add(tuple(swap))
+
+        return candidate_swaps
+
+    def _score_heuristic(self, heuristic, front_layer, extended_set, layout, swap_qubits=None):
+        """Return a heuristic score for a trial layout.
+
+        Assuming a trial layout has resulted from a SWAP, we now assign a cost
+        to it. The goodness of a layout is evaluated based on how viable it makes
+        the remaining virtual gates that must be applied.
+        """
+        if heuristic == 'basic':
+            return sum(self.coupling_map.distance(*[layout[q] for q in node.qargs])
+                       for node in front_layer)
+
+        elif heuristic == 'lookahead':
+            first_cost = self._score_heuristic('basic', front_layer, [], layout)
+            first_cost /= len(front_layer)
+
+            second_cost = self._score_heuristic('basic', extended_set, [], layout)
+            second_cost = 0.0 if not extended_set else second_cost / len(extended_set)
+
+            return first_cost + EXTENDED_SET_WEIGHT * second_cost
+
+        elif heuristic == 'decay':
+            return max(self.qubits_decay[swap_qubits[0]], self.qubits_decay[swap_qubits[1]]) * \
+                   self._score_heuristic('lookahead', front_layer, extended_set, layout)
+
+        else:
+            raise TranspilerError('Heuristic %s not recognized.' % heuristic)
+
+
+def _copy_circuit_metadata(source_dag):
```

Comment:
from the description, I suggest the name `_copy_dag_metadata`

Reply:
i copied this from the lookahead_swap pass and dag and circuit are basically the same thing